x/stakeibc/keeper/callbacks.go

```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
 
 	return nil
 }
+
+// DelegationCallback is a callback handler for Querydelegation queries.
+func DelegationCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
+	// NOTE(TEST-112) for now, to get proofs in your ICQs, you need to query the entire store on the host zone! e.g. "store/bank/key"
+
+	zone, found := k.GetHostZone(ctx, query.GetChainId())
+	if !found {
+		return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+	}
+
+	// ensure ICQ can be issued now! else fail the callback
+	valid, err := k.IsWithinICQEpochWindow(ctx)
+	if err != nil {
+		return err
+	} else if !valid {
+		return nil
+	}
+
+	qdel := stakingtypes.Delegation{}
+	err = k.cdc.Unmarshal(args, &qdel)
+	if err != nil {
+		k.Logger(ctx).Error("unable to unmarshal qdel info for zone", "zone", zone.ChainId, "err", err)
+		return err
+	}
+	k.Logger(ctx).Info(fmt.Sprintf("DelegationCallback: zone %s qdel %v", zone.ChainId, qdel))
+
+	// get tokens using the validator's conversion rate
+	for i, v := range zone.Validators {
+		k.Logger(ctx).Info(fmt.Sprintf("DELCB %s", v.Address))
+		if v.Address == qdel.ValidatorAddress {
+			// convert shares to tokens using the exchange rate
+			// TODO: make sure conversion math precision matches the sdk's staking module's version (we did it slightly differently)
+			// note: truncateInt per https://github.com/cosmos/cosmos-sdk/blob/cb31043d35bad90c4daa923bb109f38fd092feda/x/staking/types/validator.go#L431
+			qNumTokens := qdel.Shares.Mul(v.TokensFromShares).TruncateInt()
+			k.Logger(ctx).Info(fmt.Sprintf("DelegationCallback: zone %s validator %s prevNtokens %v qNumTokens %v", zone.ChainId, v.Address, v.DelegationAmt, qNumTokens))
+			if qNumTokens.Uint64() != v.DelegationAmt {
```

Comment:
Below we're using `slashAmt` - can this code only be executed when a slash occurs? Should we check that `qNumTokens.Uint64() < v.DelegationAmt`?

nit: in the check, should we add some precision factor? e.g. `qNumTokens.Uint64() < (v.DelegationAmt - epsilon)`

Reply:
Good catch on the LT check. On precision, I don't think so. These are already integers and any epsilon GT 0 would cause problems if tokens appreciate to high USD values (e.g. you don't want to abort the slash update if the slash is 1 token when that token is worth USD1K, for example)